File: src/main/java/algo/interval/InsertInterval_57.java
package algo.interval;

import java.util.ArrayList;
import java.util.Arrays;

public class InsertInterval_57 {
  public static int[][] insert(int[][] intervals, int[] newInterval) {
    var merged = new ArrayList<int[]>();
    int len = intervals.length, start = newInterval[0], end = newInterval[1];
    var result = new int[]{start, end};

    if (len == 0) {
      return new int[][]{{newInterval[0], newInterval[1]}};
    }

    var pos = 0;
    while (pos < len && intervals[pos][0] < start) {
      if (intervals[pos][1] >= start) {
        break;
      }
      merged.add(intervals[pos]);
      pos++;
    }

    // merge intervals[pos] to current
    if (pos < len && intervals[pos][0] <= end) {
      result[0] = Math.min(result[0], intervals[pos][0]);
      result[1] = Math.max(result[1], intervals[pos][1]);
      pos++;
      if (pos < len) {
        if (intervals[pos][0] <= end) {
          while (pos < len && intervals[pos][0] <= end) {
            pos++;
          }
        }
        result[1] = Math.max(result[1], intervals[pos - 1][1]);
      }
    }
    merged.add(result);

    while (pos < len) {
      merged.add(intervals[pos]);
      pos++;
    }

    var arr = new int[merged.size()][];
    return merged.toArray(arr);
  }

  public static void main(String[] args) {
    System.out.println(Arrays.deepToString(insert(new int[][]{{1, 3}, {6, 9}}, new int[]{2, 5})));
    System.out.println(Arrays.deepToString(insert(new int[][]{}, new int[]{2, 5})));
    System.out.println(Arrays.deepToString(insert(new int[][]{{1,5}}, new int[]{2, 7})));
    System.out.println(Arrays.deepToString(insert(new int[][]{{1,2},{3,5},{6,7},{8,10},{12,16}}, new int[]{4,8})));
  }
}

File: src/main/kotlin/algo/interval/InsertInterval_57.kt
package algo.interval

class KtInsertInterval_57 {
    fun insert(intervals: Array<IntArray>, newInterval: IntArray): Array<IntArray> {
        val merged: MutableList<IntArray> = mutableListOf()
        var (newStart, newEnd) = newInterval
        var startIndex = -1; var endIndex = -1
        run loop@ {
            intervals.indices.forEach { i ->
                val interval = intervals[i]
                val (start, end) = interval
                if (start < newStart) {
                    merged.add(interval)
                    startIndex = i
                }
                if (end >= newEnd && endIndex == -1) {
                    endIndex = i
                    return@loop
                }
            }
        }

        if (startIndex != -1) {
            val startInterval = intervals[startIndex]
            // merge with previous interval
            if (newStart <= startInterval[1]) {
                merged.removeLast()
                newStart = startInterval[0]
            }
        }

        if (endIndex != -1) {
            val endInterval = intervals[endIndex]
            // merge with next interval
            if (newEnd >= endInterval[0]) {
                newEnd = endInterval[1]
                endIndex++
            }
        }

        merged.add(intArrayOf(newStart, newEnd))

        if (endIndex != -1) {
            for (i in endIndex..<intervals.size) {
                merged.add(intervals[i])
            }
        }

        return merged.toTypedArray()
    }
}

fun main() {
    val insert = KtInsertInterval_57()
    println(insert.insert(arrayOf(intArrayOf(1,3), intArrayOf(6,9)), intArrayOf(2,5)))
}